Switch command line parsing to case-app.

jacarey opened the pull request:
Changing our command parsing to use case-app. This allows us to directly parse command line args into our transfer model and generally makes things much cleaner/clearer (IMO)

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: Ensure that you have added or updated tests
- [ ] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

jacarey committed:
Switch command line parsing to case-app.

jacarey commented:
This is a bit of a WIP but I wanted to get early feedback.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -1,14 +1,25 @@
-object ClioClient extends App {
+object ClioClient extends CommandAppWithPreCommand[CommonOptions, Command] {
+
+  //added to stop Intellij from removing import
```
😬  ouch

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,84 @@
+object CustomArgParsers {
+  implicit def locationParser: ArgParser[Location] = {
+    ArgParser.instance[Location]("location") { location =>
+      Right(Location.pathMatcher(location))
```
If `location` isn't a valid stringified `Location`, will this become a `Left[SomeException]` when `Location.pathMatcher(location)` throws?

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,84 @@
+    ArgParser.instance[DocumentStatus]("documentStatus") { documentStatus =>
+      Right(
+        DocumentStatus.namesToValuesMap
+          .getOrElse(documentStatus, DocumentStatus.Normal)
```
Do we want this `getOrElse` here? I think it'll make all invalid documentStatus arguments map to Normal, instead of raising an exception.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,84 @@
+  implicit def documentStatusParser: ArgParser[DocumentStatus] = {
+    ArgParser.instance[DocumentStatus]("documentStatus") { documentStatus =>
+      Right(
+        DocumentStatus.namesToValuesMap
```
Rather than having two separate parsers for `Location` and `DocumentStatus`, could we have a single `ArgParser[T <: EnumEntry]` that could handle both (and any others we might add)?

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -1,14 +1,25 @@
 
-object ClioClient extends App {
+object ClioClient extends CommandAppWithPreCommand[CommonOptions, Command] {
+
```
I think we lost the version-reporting. You should be able to `override val appVersion = <however we read the version>`

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,84 @@
+  }
+  implicit def offsetDateTimeParser: ArgParser[OffsetDateTime] = {
+    ArgParser.instance[OffsetDateTime]("date") { offsetDateAndTime =>
+      Right(OffsetDateTime.parse(offsetDateAndTime))
```
Same here, if parsing fails does this become a `Left`?

jacarey replied:
Yup just added positive cases to see if I could get it working.

jacarey committed:
Clean up

jacarey commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,84 @@
 
+final case class QueryWgsUbam(
+  @Recurse
+  transferWgsUbamV1QueryInput: TransferWgsUbamV1QueryInput,
```
Since documentation is handled by annotations I'm not entirely sure the best way to add documentation for options coming from the transfer model

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,84 @@
-    QueryWgsUbam.toString -> QueryWgsUbam
-  )
+sealed abstract class Command extends Product with Serializable {
+  def execute(webClient: ClioWebClient, bearerToken: String)(
```
I'd still prefer to keep the `execute` definitions separated from the `Command` instance definitions, to avoid being forced by `sealed` to put all the implementations in this file.

jacarey replied:
I can move execute. Though case-app had issues if this base command wasn't sealed for some reason.. I'll poke around a bit.

danxmoran reviewed (commented):
Am a fan of the overall direction, I think it'll save us a lot of boilerplate in the long run 👍 

Did you notice any slowdown in compile times?

pshapiro4broad commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -18,20 +29,27 @@ object ClioClient extends App {
     ClioClientConfig.ClioServer.clioServerPort,
```
Why is `ClioWebClient` created here instead of in the `ClioClient` constructor below?

jacarey replied:
So that we can pass in mock web clients for testing.

jacarey committed:
Handle parsing exceptions of custom types.

jacarey committed:
Move execution out of commands.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -18,20 +28,28 @@ object ClioClient extends App {
-  val client = new ClioClient(webClient)
 
-  System.exit(client.execute(args))
+  override def run(command: CommandType, remainingArgs: RemainingArgs): Unit = {
```
Should we check that `remainingArgs` is empty here?

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -18,20 +28,28 @@ object ClioClient extends App {
+    )
+  }
+
+  override def beforeCommand(options: CommonOptions,
```
Nitpick: it might be clearer how things work to have this method defined right under the `var commonOptions = _` line

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/CommandDispatch.scala:
```diff
@@ -3,31 +3,30 @@ package org.broadinstitute.clio.client.commands
     command match {
-      case AddWgsUbam   => AddWgsUbamCommand.execute(webClient, config)
-      case QueryWgsUbam => QueryWgsUbamCommand.execute(webClient, config)
+      case commandType: AddWgsUbam =>
```
Naming nitpick: rather than `commandType`, could it be `add` and `query`?

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,54 @@
+
+object CustomArgParsers {
+  implicit def locationParser: ArgParser[Location] = {
+    ArgParser.instance[Location]("location") { location =>
```
Same question as before: rather than having separate parsers for every `EnumEntry` type, could we have one general `ArgParser[T <: EnumEntry]`?

jacarey replied:
We would still need to have an instance of check in order to know which enum to query. All we get from the arg parser is the string. I can try it though and you can let me know what you think.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,30 +1,54 @@
-  ): Future[HttpResponse]
-}
+final case class QueryWgsUbam(
+                               @Recurse
```
scalafmt why

danxmoran reviewed (commented):
I really like it :tada: though I notice when I run it locally (through SBT) without any arguments it seems to hang

pshapiro4broad commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/CommandDispatch.scala:
```diff
@@ -3,31 +3,30 @@ package org.broadinstitute.clio.client.commands
+      case commandType: AddWgsUbam =>
+        new AddWgsUbamExecutor(commandType)
+          .execute(webClient, bearerToken)
+      case commandType: QueryWgsUbam =>
```
Is a way to make `CommandType` return the correct constructor to call, to avoid using `match` here? Or maybe we aren't trying to avoid using match in our scala code?

jacarey replied:
It was suggested that all execute code is handled by the dispatcher since CommandType is a sealed trait and we didn't want the commands themselves to get too big.

pshapiro4broad replied:
OK, I remember from the scala class that using `match` for `trait`s was the scala way. Am not sure I agree that it's a good way...

pshapiro4broad commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Executor.scala:
```diff
@@ -0,0 +1,12 @@
+import scala.concurrent.{ExecutionContext, Future}
+
+trait Executor {
+  def execute(webClient: ClioWebClient, bearerToken: String)(
```
Could we have a more restricted type for `bearerToken`? It seems like we don't need to perform string operations on it between its creation and its usage.

Is there a way to use an implicit to avoid having to pass it around? It will be the same for all method calls made during the lifetime of JVM, correct?

jacarey replied:
yup

pshapiro4broad commented on clio-client/src/test/resources/testdata/metadata:
```diff
@@ -1,33 +1,33 @@
-    "ubamPath" : "ubamPath",
-    "ubamSize" : 1
+  "analysis_type" : "analysisType",
+  "bait_intervals" : "analysisType",
```
Before the values were different for each field, now they're all the same. Why was this changed?

jacarey replied:
I think this had to do with how the json was being unmarshalled. I was getting leftover keys on any value that had camel case instead of snake case

pshapiro4broad reviewed (approved):
A few comments but looks like an improvement overall.

It seems like we should have at least one test that verifies that command line arguments are interpreted correctly (vs tests that create the structure that contains parsed argument data). Is the intent that the integration tests would be the place for that?

jacarey committed:
Review changes.
Add generic enum type parser.
Added system exit spec and ability to inject web client into client object so we can test arg parsing without need integration test.

jacarey commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -39,16 +45,18 @@ object CustomArgParsers {
 final case class QueryWgsUbam(
-                               @Recurse
-                               transferWgsUbamV1QueryInput: TransferWgsUbamV1QueryInput,
-                             ) extends CommandType
```
scalafmt can't decide what it wants to do here.

jacarey commented:
not in love with how we are doing DI for testing... any suggestions?

jacarey committed:
Merge branch 'develop' into jc_test_case_app

jacarey committed:
Merge branch 'develop' into jc_test_case_app

jacarey committed:
Fix up exiting tests.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -1,47 +1,74 @@
   sys.addShutdownHook({ val _ = system.terminate() })
 
-  val webClient = new ClioWebClient(
+  var webClient: ClioWebClient = new ClioWebClient(
```
Accidental `val` -> `var`?

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -1,47 +1,74 @@
+
+  //not sure this is the best way to pass common opts but case-app docs are spotty at best
+  var commonOptions: CommonOptions = _
+  UUIDParser
```
Needing this makes me sad...

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,33 +1,79 @@
-    implicit ec: ExecutionContext,
-  ): Future[HttpResponse]
-}
+final case class CommonOptions(
```
Could weeeeee move these class / type definitions into the `CustomArgParsers` object (and maybe rename it to be more generic)? Then when we import `CustomArgParsers._` and use `CommandType`, the import will be marked as used and we won't need the weird no-op `UUIDParser` statement.

danxmoran commented on project/src/main/scala/org/broadinstitute/clio/sbt/Dependencies.scala:
```diff
@@ -77,6 +78,8 @@ object Dependencies {
   val ClientMainDependencies: Seq[ModuleID] = Seq(
     "ch.qos.logback" % "logback-classic" % LogbackClassicVersion,
+    "com.github.alexarchambault" %% "case-app" % CaseAppVersion,
+    "com.iheart" %% "ficus" % FicusVersion,
```
Nitpick: can you move this below `scopt` to keep the ABC order?

jacarey replied:
scopt shouldn't be needed any more

jacarey replied:
also g comes before i

danxmoran replied:
Yup was asking to move the `com.iheart` below, but moot point with scopt going away

jacarey committed:
Moved common options to custom arg parser and remove unused implicit import.

jacarey committed:
fix val -> var
remove scopt

jacarey committed:
Codecy fixes.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -1,47 +1,69 @@
   }
 }
 
 class ClioClient(commandDispatch: CommandDispatch) {
```
Not much point to this class now.

jacarey replied:
aside from giving tests a call execute without working about sys.exit I agree. I'll try to factor out the tests that call execute directly.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala:
```diff
@@ -1,33 +1,80 @@
+  }
 }
+
+sealed trait CommandType
```
Should the rest of these be moved into `Commands` to mirror the move of `CommonOptions`?

jacarey replied:
I tried that and case-app got mad... I'll see if I can fiddle around with it more

danxmoran commented on clio-client/src/test/resources/testdata/metadata:
```diff
@@ -1,33 +1,33 @@
-    "ubam_path" : "gs://testProject/testSample/ubamPath1.unmapped.bam",
-    "ubam_size" : 1
+  "analysis_type" : "analysisType",
+  "bait_intervals" : "analysisType",
```
This might have been asked in an outdated comment, but were these all set to the same value on purpose?

jacarey replied:
Nope. Copy/paste... I'll set them to more relevant values.

danxmoran reviewed (approved):
Some final questions, but :shipit: 🎉  overall

tbl3rd commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -1,47 +1,69 @@
+    }
+  }
+
   implicit val system: ActorSystem = ActorSystem("clio-client")
```
ActorSystem(progName) ?

jacarey committed:
Review changes. Remove client spec.. didn't do much and required sys.exit catch.

jacarey committed:
Merge branch 'develop' into jc_test_case_app

# Conflicts:
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/AddWgsUbamCommand.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/CommandDispatch.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/MoveWgsUbamExecutor.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/QueryWgsUbamCommand.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/parser/BaseArgs.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala
#	clio-client/src/test/scala/org/broadinstitute/client/commands/MoveWgsUbamSpec.scala

jacarey committed:
Merge branch 'develop' into jc_test_case_app

# Conflicts:
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/AddWgsUbamCommand.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/CommandDispatch.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/Commands.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/MoveWgsUbamExecutor.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/commands/QueryWgsUbamCommand.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/parser/BaseArgs.scala
#	clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala
#	clio-client/src/test/scala/org/broadinstitute/client/commands/MoveWgsUbamSpec.scala

tbl3rd commented on clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala:
```diff
@@ -1,47 +1,54 @@
-  val commandDispatch = new CommandDispatch(webClient, IoUtil)
 
-  val client = new ClioClient(commandDispatch)
+  override def run(command: CommandType, remainingArgs: RemainingArgs): Unit = {
```
Weird. Where is remainingArgs used, and why doesn't the compiler complain?

jacarey replied:
used now.. but it definitely didn't complain

jacarey committed:
Check remaining args for subcommands.

jacarey committed:
Patch from Tom to fix command line compile issues.

jacarey committed:
Quick yaml fix.

tbl3rd commented:
FWIW: I QA this here. https://github.com/broadinstitute/tbl/blob/master/qa/qa-clio-PR-84-jc.org
Posting in the PR because I could not find a Jira ticket.
Oh, I guess this is it. https://broadinstitute.atlassian.net/browse/DSDEGP-1517
Never mind.
